src/gravel/controllers/nodes.py

```diff
@@ -200,32 +228,45 @@ def start(self):
         self._load()
         self._incoming_task = asyncio.create_task(self._incoming_msg_task())
 
-    async def join(self, address: str, token: str) -> bool:
-        logger.debug(f"=> mgr -- join > with leader {address}, token: {token}")
-
-        if self._state:
-            assert self._state.role == NodeRoleEnum.LEADER or \
-                   self._state.role == NodeRoleEnum.FOLLOWER
-            raise NodeHasJoinedError()
+    async def join(self, leader_address: str, token: str) -> bool:
+        logger.debug(
+            f"=> mgr -- join > with leader {leader_address}, token: {token}"
+        )
 
-        deployment_stage = gstate.config.deployment_state.stage
-        if deployment_stage != DeploymentStage.none:
+        assert self._state
+        if self._state.stage == NodeStageEnum.BOOTSTRAPPING:
+            raise NodeBootstrappingError()
+        elif self._state.stage == NodeStageEnum.BOOTSTRAPPED:
             raise NodeHasBeenDeployedError()
+        elif self._state.stage == NodeStageEnum.JOINING:
+            raise NodeAlreadyJoiningError()
+        elif self._state.stage == NodeStageEnum.READY:
+            raise NodeHasJoinedError()
```

Comment:
(nit) this set of checks could be simplified.. eg check if it's not in desired state and raise an unexpected state error that logs current state.